pkg/kn/core/root_test.go

```diff
@@ -40,6 +41,69 @@ func TestNewDefaultKnCommand(t *testing.T) {
 	})
 }
 
+func TestNewDefKnCmdWithArgsforCmdValiditySuccess(t *testing.T) {
+
+	pluginHandler := plugin.NewDefaultPluginHandler(plugin.ValidPluginFilenamePrefixes,
+		"/random/plugins", false)
+	args := []string{"kn", "service", "update", "--tag=0.13"}
+	rootCmd := NewDefaultKnCommandWithArgs(NewKnCommand(), pluginHandler, args, os.Stdin, os.Stdout, os.Stderr)
+	checkRootCmd(t, rootCmd)
+}
+
+func TestNewDefKnCmdWithArgsforWrongCmd(t *testing.T) {
+
+	if os.Getenv("EXIT") == "1" {
+		pluginHandler := plugin.NewDefaultPluginHandler(plugin.ValidPluginFilenamePrefixes,
+			"/random/plugins", false)
+		args := []string{"kn", "srevice", "update", "--tag=0.13"}
+		NewDefaultKnCommandWithArgs(NewKnCommand(), pluginHandler, args, os.Stdin, os.Stdout, os.Stderr)
+		return
+	}
+	cmd := exec.Command(os.Args[0], "-test.run=TestNewDefKnCmdWithArgsforWrongCmd")
+	cmd.Env = append(os.Environ(), "EXIT=1")
+	err := cmd.Run()
+	if e, ok := err.(*exec.ExitError); ok && !e.Success() {
+		return
+	}
+	t.Fatalf("process ran with err %v, want exit status 1", err)
+}
+
+func TestNewDefKnCmdWithArgsforWrongVerb(t *testing.T) {
+
+	if os.Getenv("EXIT") == "1" {
+		pluginHandler := plugin.NewDefaultPluginHandler(plugin.ValidPluginFilenamePrefixes,
+			"/random/plugins", false)
+		args := []string{"kn", "revision", "update", "--tag=0.13"}
+		NewDefaultKnCommandWithArgs(NewKnCommand(), pluginHandler, args, os.Stdin, os.Stdout, os.Stderr)
+		return
+	}
+	cmd := exec.Command(os.Args[0], "-test.run=TestNewDefKnCmdWithArgsforWrongVerb")
```

Comment:
Nice trick to call the same test again from the outside, but I wonder whether we couldn't make it  a real unit test. This currently looks more like an integration test which should be triggered from the outsied (e.g. including it into our e2e test suite which execs to kn).

So I would suggest to call `NewDefaultKnCommndWithArgs` directly within the test, with the arguments prepared  accordingly. This is then would be a real self-container minmal unit -test (which does not have any deps on the external context, e.g. where the test is run). This is faster and more robust. Let's leave integration tests to teh e2e test suite (you could add such a test there, too)

Reply:
Thank you @rhuss .  The reason I had to do this was the `NewDefaultKnCommndWithArgs` func exits with os.Exit(1) when there is an error. So for failure cases the test exits without continuing. Is there something I could do for this scenaro ?

Reply:
@rhuss moved the tests to e2e.